### PR TITLE
Introducing the XGBoost vector leaf model.

### PR DESCRIPTION
**Pull request by trivialfis**

### CODE REVIEW

File: _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
+---
+layout: post
+title: "Introducing the XGBoost Vector-Leaf Model"
+date:   2026-8-25
+author: Jiaming Yuan and Rory Mitchell
+categories:
+comments: true
+---
+
+Most decision trees store a single number in each leaf. This works naturally for ordinary
```
Comment (mgrauer) [inline _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md]: number => output or target?

File: _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+regression, where a model predicts one value for every row. Many problems, however,
+require several related predictions at once: multiclass classification, quantile and
+expectile regression, and multi-output regression are common examples.
+
+By default, XGBoost handles these problems by building one stack of scalar-leaf trees for
+each output. The vector-leaf model offers another choice: build one shared tree whose
+leaves hold a vector of predictions. In this post, we explain the trade-off and explore it
+across a range of multiclass classification datasets.
+
```
Comment (mgrauer) [inline _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md]: it'd be good to hoist the value to the user to the intro, repeating the info from below. allow people to see *why* they should care in the intro section.

you can use some content along these lines that you already have:

>A vector-leaf tree asks the question once, then stores a different adjustment for every class in the resulting leaf.

>when one partition of the data works as well as many separate partitions, prefer the simpler representation. It can be smaller and faster to evaluate, and the shared structure can act as a useful modeling constraint, making the model more likely to generalize well for unseen data.

File: _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+separate score update for every class, but all classes follow the same sequence of splits
+that routes a row to that leaf. For a problem with 32 classes, one scalar boosting round
+adds 32 trees, while one vector boosting round adds a single shared tree. This difference
+can produce far fewer splits and a much simpler model.
+
+![Three scalar-leaf class trees compared with one shared vector-leaf tree](/images/blog/2026-08-25-introducing-the-xgboost-vector-leaf-model/scalar_vs_vector_leaf_trees.png)
+
+## Why share a tree?
+
+Suppose a tree asks whether a pixel is dark, a playing card is an ace, or a sensor reading
```
Comment (mgrauer) [inline _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md]: nit

"a tree asks" => is tree the right subject/verb here, or maybe "model answers" ? 

I see a little bit lower that "a vector-leaf tree asks the question once" so maybe you are exactly right already without any changes. if you do you make a change in one sentence, change it in both for consistency.

File: _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+exceeds a threshold. In a multiclass problem, the answer may be useful to several classes
+at once. A vector-leaf tree asks the question once, then stores a different adjustment for
+every class in the resulting leaf.
+
+This is an appealing example of Occam's razor: when one partition of the data works as
+well as many separate partitions, prefer the simpler representation. It can be smaller and
+faster to evaluate, and the shared structure can act as a useful modeling constraint,
+making the model more likely to generalize well for unseen data.
+
+Sharing is also the central constraint. Vector leaves work best when outputs benefit from
```
Comment (mgrauer) [inline _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md]: These two paragraphs are good, I just didn't follow exactly what you are trying to say with "Sharing is also the central constraint.", why you thought that was important, and how that related to the rest of the content in these two paragraphs.

File: _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+problem with 32 normally distributed features. Every target combines those same set of
+features with different weights and a small amount of random noise. The targets therefore
+need different predictions, but they all benefit from the same basic split questions. We
+generated `2^20` training rows and `2^18` test rows, then compared problems with 2, 4, 8,
+16, and 32 outputs. Both strategies were trained for 128 boosting rounds. This
+deliberately partition-compatible dataset is the favorable case for vector leaves.
+
+![GPU scaling with compatible outputs. Error bars show variation across three timing repeats; the generated data are fixed.](/images/blog/2026-08-25-introducing-the-xgboost-vector-leaf-model/computational_efficiency_gpu.png)
+
+Both strategies retain essentially the same predictive error. Their complexity, however,
```
Comment (mgrauer) [inline _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md]: you don't mention anything in this section or the images about what hardware you evaluated this on, and whether there is a meaningful gpu vs cpu difference.

that may not be an important part of the story. 

i do see though in the alt text that you mention GPUs, but not anywhere in the paragraph text

File: _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+## Why share a tree?
+
+Suppose a tree asks whether a pixel is dark, a playing card is an ace, or a sensor reading
+exceeds a threshold. In a multiclass problem, the answer may be useful to several classes
+at once. A vector-leaf tree asks the question once, then stores a different adjustment for
+every class in the resulting leaf.
+
+This is an appealing example of Occam's razor: when one partition of the data works as
+well as many separate partitions, prefer the simpler representation. It can be smaller and
+faster to evaluate, and the shared structure can act as a useful modeling constraint,
```
Comment (mgrauer) [inline _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md]: not just faster to evaluate but also faster to train! is that worth adding?

Comment (mgrauer) [reply _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md]: ok maybe not. after reading the "what happens on real data" section i see that the training time is not always favored for vector leaf

File: _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+Vector leaves achieve lower held-out loss on nine of ten datasets and use fewer splits and
+less disk space on all ten. Poker Hand is the standout, but the pattern also appears on
+balanced datasets such as EMNIST and Devnagari, so class imbalance does not explain the
+result by itself.
+
+Some gains come with longer training. On Devnagari, the vector model uses about three
+times as many boosting rounds to train. It nevertheless finishes substantially more
+accurate and about one third the size. Letter Recognition improves both predictive quality
+and training time, while Gas Sensor Drift retains a smaller advantage across a
+chronological shift. Shared structure is an advantage when it preserves enough predictive
```
Comment (mgrauer) [inline _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md]: i'm unclear on what "chronological shift" means. i would naively assume that means shifting a start or end time. if it means greater training time, that is strange wording IMO

File: _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+flexibility.
+
+## A practical default for multiclass classification
+
+These results make vector leaves a strong starting point for multiclass classification,
+especially as the number of classes grows. Rather than treating `multi_output_tree` only
```
Comment (mgrauer) [inline _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md]: given that you say "especially as the number of classes grows.", would it make sense to explain in the comparison chart above between all of the test datasets how they rank against each other with resect to number of classes? are they ordered in that chart by number of classes?

File: _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+as a later optimization, it is reasonable to try it first and retain a scalar model as the
+comparison.
+
+Give the vector model more boosting rounds. A scalar round adds one tree for every
```
Comment (mgrauer) [inline _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md]: "Give the vector model more boosting rounds." seems like a non sequeter to me. how does this sentence fit in with the above paragraph or the current paragraph?

Comment (mgrauer) [reply _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md]: ok reading the rest of this paragraph i think the point of including this might be to say that compared to scalar models, vector models need more boosting rounds. if that is true you can clarify the sentence

File: _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+class, whereas a vector round adds only one shared tree; equal round limits therefore give
+the scalar model much more tree structure. On K49 and Devnagari, the vector models
+continued improving for substantially longer and eventually became both more accurate and
+smaller than their scalar counterparts.
+
+A lower learning rate is also worth pairing with the larger round cap. XGBoost's default
```
Comment (mgrauer) [inline _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md]: nit

"A lower learning rate is also worth pairing with the larger round cap"

might be good to be a bit more explicit, e.g.

A lower learning rate is also worth pairing with the larger round cap when training vector leaf models.

though i admit, it is somewhat implicit

File: _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+learning rate is quite large; reducing it makes each tree a smaller update and usually
+enables more boosting rounds. The smaller steps can make the shared, multidimensional
```
Comment (mgrauer) [inline _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md]: nit

i agree with copilot about run-on sentences. at least don't have two semi-colon sentences in a row, that is a little distracting.

File: _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+updates easier to optimize; Dionis, for example, improved after its vector learning rate
+was reduced. Early stopping can then select the useful number of rounds without committing
+to the full cap.
+
+## About the experiments
+
+Every real-data run used GPU histogram training and `QuantileDMatrix` with a DGX Spark.
```
Comment (mgrauer) [inline _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md]: this section probably answers my questions above about what hardware and also chronological shift. not that i understand chronological shift, but it seems to be something related to Gas Sensor Drift dataset/modeling.

**Review by mgrauer — CHANGES_REQUESTED** [review]
This blog looks great! Thank you for putting it together.

**Comment by mgrauer** [comment]
I recommend reading all the way through my review comments before attempting to address any of them, because I may have answered some of the questions in a later section that I raised in an earlier section.

File: _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md (hunk 2 above)
Comment (trivialfis) [reply _posts/2026-08-25-introducing-the-xgboost-vector-leaf-model.md]: Mentioned the advantage in the intro section.

**Review by RAMitchell — APPROVED** [review]
Very good! The technical level is right, the results are very clearly presented and the conclusions clear.

**Comment by trivialfis** [comment]
Thank you both for the reviews!
